### PR TITLE
Service start order

### PR DESCRIPTION
**Pull request by antmat**
Currently service start order matters and it only happens to work as locator goes first in alphabetical order.
This PR does the following:
  * Ensures all service start signals are issued AFTER all services (including locator) have started.
  * Starts locator after all other services. In this case it can use started services for it's purposes.
  * Fix segmentation fault in case "locate" could not find actor in map

### CODE REVIEW

File: include/cocaine/context.hpp [hunk 1]
```diff
@@ -90,8 +90,25 @@ struct config_t {
     };
 
     typedef std::map<std::string, component_t> component_map_t;
+    typedef std::vector<std::pair<std::string, component_t>> components_t;
+
+    struct component_order_t {
+        bool
+        operator()(const components_t::value_type& c1, const components_t::value_type& c2) const {
```
Comment (3Hren) [inline include/cocaine/context.hpp]: `lhs/rhs` are more convenient.

File: include/cocaine/context.hpp [hunk 2]
```diff
@@ -90,8 +90,25 @@ struct config_t {
     };
 
     typedef std::map<std::string, component_t> component_map_t;
+    typedef std::vector<std::pair<std::string, component_t>> components_t;
```
Comment (Kobolog) [inline include/cocaine/context.hpp]: Why not an `std::map<K, V, CustomCompare>`?

File: include/cocaine/context.hpp [hunk 3]
```diff
@@ -152,6 +169,9 @@ class context_t {
     // because services are allowed to start and stop other services during their lifetime.
     synchronized<service_list_t> m_services;
 
+    //Flag set after initialization
+    bool ready;
```
Comment (Kobolog) [inline include/cocaine/context.hpp]: In my opinion, boolean flags are usually a manifestation of a bad design. I'd rather accept somewhat larger refactoring, but not a bool flag crutch.

File: src/context.cpp [hunk 4]
```diff
@@ -582,26 +585,26 @@ context_t::insert(const std::string& name, std::unique_ptr<actor_t> service) {
...
+        if (std::count_if(ptr->begin(), ptr->end(), match{name})) {
             throw cocaine::error_t("service '%s' already exists", name);
         }
 
         service->run();
 
         COCAINE_LOG_INFO(m_logger, "service has been started")(
-            "service", name
-        );
+                "service", name
```
Comment (Kobolog) [inline src/context.cpp]: Some weird whitespace.

File: src/context.cpp [hunk 5]
```diff
@@ -640,9 +643,11 @@ boost::optional<const actor_t&>
 context_t::locate(const std::string& name) const {
     auto ptr = m_services.synchronize();
     auto it  = std::find_if(ptr->begin(), ptr->end(), match{name});
-
+    if(it == ptr->end()) {
+        return boost::none;
+    }
     return boost::optional<const actor_t&>(
-        it != ptr->end() && it->second->is_active(),
```
Comment (Kobolog) [inline src/context.cpp]: I think a race condition does still exist here.

File: include/cocaine/context.hpp (hunk 2 above)
Comment (antmat) [inline include/cocaine/context.hpp]: Because it would be compared by value, not by key.

File: src/context.cpp (hunk 5 above)
Comment (antmat) [inline src/context.cpp]: Only in terms that reference can become dangling one if someone erases actor from internal storage. We can pointer type from uniqe to shared and return weak_ptr in locate. Is it suitable?

File: include/cocaine/context.hpp (hunk 3 above)
Comment (antmat) [inline include/cocaine/context.hpp]: In this case I can try to write some class similar to boost::signals (or even a wrapper on boost signals) which tracks all invocations and apply them to newly attached slots.
Or we can stop using signals and move that functions to basic_dispatch interface. In this case usage will be limited only by basic_dispatch derived classes. I don't really like both of them - maybe you have a better idea?

Comment (Kobolog) [inline include/cocaine/context.hpp]: Is it possible to add a new behavior to `boost::signals2` so that the slot could be blocked in a deferred sort of state where it accumulates all the incoming signals but triggers them only when it's unblocked again?

File: src/context.cpp (hunk 5 above)
Comment (Kobolog) [inline src/context.cpp]: Yeah, that's what I meant — if something calls `locate()`, stores the reference somewhere and then some other codepath removes the service, the reference will be invalid. It was never an issue though, but still is an UB. It might be fixed with `weak_ptr<T>`, but I'm not sure if it won't break anything else, can you investigate? As for this PR, I'd suggest dropping that part because it's irrelevant here and moving it to a new PR.

File: src/context.cpp [hunk 6]
```diff
@@ -582,26 +585,26 @@ context_t::insert(const std::string& name, std::unique_ptr<actor_t> service) {
         logging::keyword::source() = "core"
     }));
 
-    const actor_t& actor = *service;
-
+    bool expose;
+    auto& actor = *service;
```
Comment (Kobolog) [inline src/context.cpp]: Lost `const` here.

File: src/context.cpp [hunk 7]
```diff
@@ -582,26 +585,26 @@ context_t::insert(const std::string& name, std::unique_ptr<actor_t> service) {
...
     {
         auto ptr = m_services.synchronize();
-
-        if(std::count_if(ptr->begin(), ptr->end(), match{name})) {
+        expose = ready;
+        if (std::count_if(ptr->begin(), ptr->end(), match{name})) {
```
Comment (Kobolog) [inline src/context.cpp]: Whitespace after `if`.

File: src/context.cpp (hunk 5 above)
Comment (Kobolog) [inline src/context.cpp]: Also, what was the point of breaking `it != ptr->end() && it->second->is_active()` into 2 parts? How does that fix a segmentation fault?

Comment (antmat) [inline src/context.cpp]: Because iterator to end dereference could took place. Now it can not.

File: include/cocaine/context.hpp (hunk 3 above)
Comment (antmat) [inline include/cocaine/context.hpp]: As I see from docs - there is no such possibility right out of the box - but I think I can try to implement one. It will get some time though.

File: src/context.cpp (hunk 5 above)
Comment (Kobolog) [inline src/context.cpp]: How could that happen, exactly?

File: include/cocaine/context.hpp (hunk 3 above)
Comment (Kobolog) [inline include/cocaine/context.hpp]: It should be possible to implement a deferred stateful combiner, I guess.

File: src/context.cpp (hunk 5 above)
Comment (antmat) [inline src/context.cpp]: In it was dereferenced before passing to boost::optional constructor even if predicate is false.

Comment (Kobolog) [inline src/context.cpp]: Oh, right.

Comment (Kobolog) [inline src/context.cpp]: Anyway, move it out to a separate PR, because it has nothing to do with service start ordering.

**Comment by antmat** [comment]
I've made some refactoring. As fix of UB in locate is connected with change of storage type from unique_ptr to shared_ptr which is connected with deferred signal implementation -  I've leaved that all in one PR.

**Comment by antmat** [comment]
I'll make a clean PR
